pkg/erofs/erofs.go

```diff
@@ -690,87 +724,155 @@ func (i *Inode) blocks() uint64 {
 //
-// All directory entries are _strictly_ recorded in alphabetical order.
-func (i *Inode) IterDirents(cb func(name string, typ uint8, nid uint64) error) error {
+// refer: https://docs.kernel.org/filesystems/erofs.html#directories
+func (i *Inode) getDirentName(d *Dirent, block blockData, lastDirent bool) ([]byte, error) {
+	var nameLen uint32
+	if lastDirent {
+		nameLen = block.size - uint32(d.NameOff)
+	} else {
+		nameLen = uint32(direntAfter(d).NameOff - d.NameOff)
+	}
+	if uint32(d.NameOff)+nameLen > block.size || nameLen > MaxNameLen || nameLen == 0 {
+		log.Warningf("Corrupted dirent at inode (nid=%v)", i.Nid())
+		return nil, linuxerr.EUCLEAN
+	}
+	name, err := i.image.BytesAt(block.base+uint64(d.NameOff), uint64(nameLen))
+	if err != nil {
+		return nil, err
+	}
+	if lastDirent {
+		// Optional padding may exist at the end of a block.
+		n := bytes.IndexByte(name, 0)
+		if n == 0 {
+			log.Warningf("Corrupted dirent at inode (nid=%v)", i.Nid())
+			return nil, linuxerr.EUCLEAN
+		}
+		if n != -1 {
+			name = name[:n]
+		}
+	}
+	return name, nil
+}
+
+// getDirent0 returns a pointer to the first dirent in the given block of this inode.
+func (i *Inode) getDirent0(block blockData) (*Dirent, error) {
+	d0, err := i.image.direntAt(block.base)
+	if err != nil {
+		return nil, err
+	}
+	// Apart from the offset of the first filename, nameOff0 also indicates
+	// the total number of dirents in this block.
+	if d0.NameOff < DirentSize || uint32(d0.NameOff) >= block.size {
+		log.Warningf("Invalid nameOff0 %v at inode (nid=%v)", d0.NameOff, i.Nid())
+		return nil, linuxerr.EUCLEAN
+	}
+	return d0, nil
+}
+
+// Lookup looks up a child by the name. The child inode number will be returned on success.
+func (i *Inode) Lookup(name string) (uint64, error) {
 	if !i.IsDir() {
-		return linuxerr.ENOTDIR
+		return 0, linuxerr.ENOTDIR
 	}
 
-	blocks := i.blocks()
-	blockSize := i.image.BlockSize()
+	// Currently (Go 1.21), there is no safe and efficient way to do three-way
+	// string comparisons, so let's convert the string to a byte slice first.
+	nameBytes := gohacks.ImmutableBytesFromString(name)
 
-	start := i.dataOff
-	if blocks == 1 && i.idataOff != 0 {
-		start = i.idataOff
+	// In EROFS, all directory entries are _strictly_ recorded in alphabetical
+	// order. The lookup is done by directly performing binary search on the
+	// disk data similar to what Linux does in fs/erofs/namei.c:erofs_namei().
+	type blockInfo struct {
+		block      blockData
+		numDirents uint16
 	}
+	var target blockInfo
 
-	// Iterate all the blocks which contain dirents.
-	for blocks > 0 {
-		// Get the first dirent in the current block.
-		direntOff := start
+	// Find the block that may contain the target dirent first.
+	left, right := int64(0), int64(i.blocks)-1
+	for left <= right {
+		// Cast to uint64 to avoid overflow.
+		mid := uint64(left+right) >> 1
```

Comment:
To prevent the potential overflow, the cast has to happen on `left` and `right` first and then the addition can happen.

Separately, `i.blocks` is a uint64 itself, why not just use uint64 for left and right?

Reply:
> To prevent the potential overflow, the cast has to happen on `left` and `right` first and then the addition can happen.

I think it's not really necessary to do the cast before performing the addition, because the binary representation of the sum result is the same. And the key point is that we should not perform a signed right shift operation on the sum result (e.g. on x86, we should use `shr` and not `sar` [1]). This is also what Go does in [`sort.Search`](https://github.com/golang/go/blob/3e67f46d4f7d661504d281bdedbd1432c09bd751/src/sort/search.go#L63).

[1] https://www.felixcloutier.com/x86/sal:sar:shl:shr

> Separately, `i.blocks` is a uint64 itself, why not just use uint64 for left and right?

Potentially, `right` can be a negative value, so `right` has to be signed. Otherwise, the `left <= right` check won't work as expected in this case.

Reply:
In reality, the sum will never lead to an int64 overflow. If we want to avoid the casts, maybe we can just rephrase the comment and remove all the unnecessary casts:

```go
	// Find the block that may contain the target dirent first.
	bLeft, bRight := int64(0), int64(i.blocks)-1
	for bLeft <= bRight {
		// The sum will never lead to an int64 overflow, as the maximum value
		// of the operands in math is ceil(MaxUint64/BlockSize).
		mid := (bLeft + bRight) >> 1
		block := i.getBlockDataInfo(uint64(mid))
		d0, err := i.getDirent0(block)
		if err != nil {
			return 0, err
		}
		numDirents := d0.NameOff / DirentSize
		name0, err := i.getDirentName(d0, block, numDirents == 1)
		if err != nil {
			return 0, err
		}
		switch bytes.Compare(nameBytes, name0) {
		case 0:
			// Found the target dirent.
			return d0.Nid(), nil
		case 1:
			// name > name0, this block may contain the target dirent.
			target = targetInfo{block, numDirents}
			bLeft = mid + 1
		case -1:
			// name < name0, this is not the block we're looking for.
			bRight = mid - 1
		}
	}
```

Reply:
I see, makes sense. We can leave it as it is then. Will leave it up to you.

Reply:
OK, I will leave it as it is.